components/form/form-directive.ts

```diff
@@ -68,108 +69,106 @@ class FormController {
     }
 
     private fetchUiSchema() {
 
-        if (this.$scope.uiSchema && this.$scope.asyncUiSchema()) {
-            throw new Error("You cannot specify both the 'ui-schema' and the 'async-ui-schema' attribute at the same time.")
-        } else if (this.$scope.uiSchema) {
-            var p = this.$q.defer();
-            p.resolve(this.$scope.uiSchema);
+        if (FormController.isUiSchemaProvider(this.scope.uiSchema)) {
+            return this.scope.uiSchema.getUiSchema();
+        } else if (typeof this.scope.uiSchema === "object") {
+            let p = this.$q.defer();
+            p.resolve(this.scope.uiSchema);
             return p;
-        } else if (this.$scope.asyncUiSchema()) {
-            return this.$scope.asyncUiSchema();
         }
 
-        // return undefined to indicate that no way of obtaining a ui schema was defined
-        // TODO: Maybe return defaultUISchema or generator function?
-        var p = this.$q.defer();
+        // if we return undefined the caller will generate a default UI schema
+        let p: ng.IDeferred<any> = this.$q.defer<any>();
         p.resolve(undefined);
         return p;
     }
 
     private fetchData() {
-        var dataProvider: JSONForms.IDataProvider = <JSONForms.IDataProvider> this.$scope.asyncDataProvider;
-        var data = this.$scope.data;
-
-        if (dataProvider && data) {
-            throw new Error("You cannot specify both the 'data' and the 'async-data-provider' attribute at the same time.")
-        } else if (dataProvider) {
-            return dataProvider.fetchData();
-        } else if (this.$scope.data) {
+        if (FormController.isDataProvider(this.scope.data)) {
+            return this.scope.data.fetchData();
+        } else if (typeof this.scope.data === "object") {
             var p = this.$q.defer();
-            p.resolve(this.$scope.data);
+            p.resolve(this.scope.data);
             return p.promise;
         }
 
         throw new Error("Either the 'data' or the 'async-data-provider' attribute must be specified.")
     }
+
+    private static isDataProvider(testMe: any): testMe is JSONForms.IDataProvider {
+        return testMe !== undefined && testMe.hasOwnProperty('fetchData');
+    }
+
+    private static isUiSchemaProvider(testMe: any): testMe is JSONForms.IUiSchemaProvider {
+        return testMe !== undefined && testMe.hasOwnProperty('fetchUiSchema');
+    }
 }
 
 interface JsonFormsDirectiveScope extends ng.IScope {
+    schema: any;
+    uiSchema: any;
+    data: any;
+}
 
-    schema: string;
-    uiSchema: string;
-    data: string;
 
-    asyncSchema: () => any;
-    asyncUiSchema: () => any;
-    asyncDataProvider: JSONForms.IDataProvider;
+class JsonFormsDirective implements ng.IDirective {
+    restrict = "E";
+    replace = true;
+    templateUrl = 'components/form/form.html';
+    controller = FormController;
+    controllerAs = 'vm';
```

Comment:
What exactly do you mean :) There are actually two things happerning here:
1. I've removed the async-* attribute variants from the `jsonforms` directive since I think this is only confusing users. Instead one can either specify the name of respective Provider or the object literal and we will dispatch accordingly.
2. Usage of `controllerAs` syntax, which binds the controller to the current scope, and make it available via `vm` in the `form.html` template.